// node-habit/index.js
var request = require('superagent'),
  util = require('util');

function HabitRpg(userId, apiKey, apiUrl) {
  this.userId = userId;
  this.apiKey = apiKey;

  this.apiUrl = apiUrl || 'https://habitrpg.com';

  this.addTask = function(type, text, optional, callback) {
    var theRequest = request.post(this.apiUrl + '/api/v1/user/task')
      .type('application/json')
      .set('Accept: gzip, deflate')
      .set('x-api-user', this.userId)
      .set('x-api-key', this.apiKey)
      .send({
        type: type,
        text: text
      });

    if (optional) {
      theRequest.send(optional);
    }
    theRequest.end(function(res) {
      if (res.ok) {
        console.log('Added: ' + text);
        console.log('Status code: ' + res.status);
        console.log('Technical details: ' + util.inspect(res.body));
        if (callback) {
          callback(res.text);
        }
      } else {
        console.log('Error in addTask: ' + util.inspect(res.text));
      }
    });
  };
}

module.exports = HabitRpg;
