// node-habit/index.js
var request = require('superagent'),
  util = require('util');

function HabitRpg(userId, apiKey, apiUrl) {
  this.userId = userId;
  this.apiKey = apiKey;

  this.apiUrl = apiUrl || 'https://habitrpg.com';

  this.addTask = function(type, text, optional, callback) {
    // We've still got requestStuff available
    // TODO: Need to inject this in the end once I abstract this out
    requestStuff.path = '/api/v1/user/task';

    habitRequestPath = requestStuff.protocol + '://' + requestStuff.host + ':' + requestStuff.port + requestStuff.path;
    console.log("HabitRPG request path: " + habitRequestPath);

    var theRequest = request.post(this.apiUrl + '/api/v1/user/tasks')
      .type('application/json')
      .set('Accept: gzip, deflate')
      .set('x-api-user', this.userId)
      .set('x-api-key', this.apiKey)
      .send({
        type: type,
        text: text
      });

    if (optional) {
      theRequest.send(optional);
    }
    theRequest.end(function(res) {
      if (res.ok) {
        console.log('Added: ' + text);
        if (callback) {
          callback(res.text);
        }
      }
      else {
        console.log('Error in addTask: ' + util.inspect(res));
      }
    });
  };
}

module.exports = HabitRpg;
